x-pack/plugins/fleet/public/applications/fleet/sections/agent_policy/create_package_policy_page/single_page_layout/hooks/setup_technology.ts

```diff
@@ -36,14 +36,15 @@ export function useSetupTechnology({
 }: {
   updateNewAgentPolicy: (policy: NewAgentPolicy) => void;
   newAgentPolicy: NewAgentPolicy;
-  updateAgentPolicy: (policy: AgentPolicy) => void;
+  updateAgentPolicy: (policy: AgentPolicy | undefined) => void;
   setSelectedPolicyTab: (tab: SelectedPolicyTab) => void;
 }) {
   const { isAgentlessEnabled } = useAgentlessPolicy();
   const [selectedSetupTechnology, setSelectedSetupTechnology] = useState<SetupTechnology>(
     SetupTechnology.AGENT_BASED
   );
   const [agentlessPolicy, setAgentlessPolicy] = useState<AgentPolicy | undefined>();
+  const [currentAgentPolicy, setCurrentAgentPolicy] = useState<AgentPolicy | undefined>();
```

Comment:
Why do we need this state? What confuses me :
1. we set it in the effect that fetches the agentless policy. But this is incorrect logically, as there is no connection between this effect and setting up the actual agent policy. It doesn't do any `updateAgentPolicy`
2. in the `handleSetupTechnologyChange` the usage of `currentAgentPolicy` is quite confusing. 
```
setCurrentAgentPolicy(agentlessPolicy);
updateAgentPolicy(currentAgentPolicy || agentlessPolicy);
```
first of all it leaves the impression that as `currentAgentPolicy` is set to `agentlessPolicy` right before calling `updateAgentPolicy` it will always be `agentlessPolicy` anyway. But mind that `setState` is async in nature, so the previous value of the `currentAgentPolicy` will be used. But as the two possible values for `currentAgentPolicy` are `agentlessPolicy` or `undefined`, this code is just equal to `updateAgentPolicy(agentlessPolicy);`

I suspect that we don't need this state at all and it should be enough to just set `updateAgentPolicy(undefined);` when setupTechnology is `AGENT_BASED`

Reply:
Unless remove the useEffect and refactor, there will be a race condition to  agent policy  reset to `agentless` policy. This essentially affect the agent count and causes the bug.

> 1. we set it in the effect that fetches the agentless policy. But this is incorrect logically, as there is no connection between this effect and setting up the actual agent policy. It doesn't do any updateAgentPolicy
>>  Since the Agentless Feature Flag is enabled and  Agentless option  is selected by default. I set the currentAgentPolicy state to agentless policy.

I added the `isDirty` check; however, the race condition continues so my next thought is to keep track of the current agent policy. The `currentAgentPolIcy` state helps track the latest agent policy when user selects setupTechnogy which can be the following `agentless`, `undefined`, and `agent-based`,  When the `handleSetupTechnologyChange`  gets passed a new value,  currentAgentPolicy has seperate state isn't affected by the useEffect race condition which keep updating agent policy to `agentless`.

>`I suspect that we don't need this state at all and it should be enough to just set updateAgentPolicy(undefined); when setupTechnology is AGENT_BASED`
>>  I tried before calling `updateAgentPolicy(undefined)` and again useEffect kept resetting agent policy id to `agentless`. I made sure to test the fix before pushing changes.

> first of all it leaves the impression that as currentAgentPolicy is set to agentlessPolicy right before calling updateAgentPolicy it will always be agentlessPolicy anyway. 
>> No that's not actual result again because we are using useEffect agentless policy keeps being reset to agentless policy.

We either refactor where  `useEffect` but that can take additional time